packages/core-mobile/app/new/contexts/SignupProvider.tsx

```diff
@@ -0,0 +1,117 @@
+import { TotpChallenge } from '@cubist-labs/cubesigner-sdk'
+import React, {
+  createContext,
+  Dispatch,
+  ReactNode,
+  SetStateAction,
+  useCallback,
+  useContext,
+  useMemo,
+  useState
+} from 'react'
+import SeedlessService from 'seedless/services/SeedlessService'
+import AnalyticsService from 'services/analytics/AnalyticsService'
+import { useRouter } from 'expo-router'
+import { copyToClipboard } from 'new/utils/clipboard'
+import { TotpErrors } from 'seedless/errors'
+import { Result } from 'types/result'
+import { hideLogoModal, showLogoModal } from 'new/components/LogoModal'
+import { OidcAuth } from 'new/types'
+
+export interface SignupContextState {
+  handleAccountVerified: () => Promise<void>
+  handleCopyCode: () => void
+  onVerifyCode: (code: string) => Promise<Result<undefined, TotpErrors>>
+  onVerifySuccess: () => void
+  totpKey?: string
+  totpChallenge?: TotpChallenge
+  setTotpChallenge: Dispatch<SetStateAction<TotpChallenge | undefined>>
+  oidcAuth?: OidcAuth
+  setOidcAuth: Dispatch<SetStateAction<OidcAuth | undefined>>
+  allowsUserToAddLater: boolean
+}
+
+export const SignupContext = createContext<SignupContextState>(
+  {} as SignupContextState
+)
+
+export const SignupProvider = ({
+  children
+}: {
+  children: ReactNode
+}): React.JSX.Element => {
+  const router = useRouter()
+  const [oidcAuth, setOidcAuth] = useState<OidcAuth>()
+  const [totpChallenge, setTotpChallenge] = useState<TotpChallenge>()
+
+  const totpKey = useMemo(() => {
+    if (totpChallenge?.totpUrl) {
+      return (
+        new URL(totpChallenge.totpUrl).searchParams.get('secret') ?? undefined
+      )
+    }
+  }, [totpChallenge])
+
+  const handleCopyCode = useCallback((): void => {
+    totpKey && copyToClipboard(totpKey, 'Code copied')
+  }, [totpKey])
+
+  const onVerifyCode = useCallback(
+    async (code: string): Promise<Result<undefined, TotpErrors>> => {
+      await totpChallenge?.answer(code)
+
+      if (oidcAuth) {
+        return SeedlessService.sessionManager.verifyCode(
+          oidcAuth.oidcToken,
+          oidcAuth.mfaId,
+          code
+        )
+      } else {
+        return { success: true, value: undefined }
+      }
+    },
+    [oidcAuth, totpChallenge]
+  )
+
+  const handleAccountVerified = useCallback(async (): Promise<void> => {
+    showLogoModal()
+    const walletName = await SeedlessService.getAccountName()
+    hideLogoModal()
+
+    if (walletName) {
+      router.navigate('./createPin')
+      return
+    }
+    router.navigate('./nameYourWallet')
+  }, [router])
+
+  const onVerifySuccess = useCallback((): void => {
+    router.dismissAll()
+    router.back()
+    handleAccountVerified()
+    AnalyticsService.capture('SeedlessMfaVerified', {
+      type: 'Authenticator'
+    })
+  }, [router, handleAccountVerified])
```

Comment:
I think it would be good to also move the handleAccountVerified function and onVerifySuccess to the screen.

Reply:
handleAccountVerified is used in multiple place during the signup flows, I think it is better to just leave it in the context, since this context is very much for signup flow anyway.

Reply:
While that’s true in terms of code reusability, I think it’s more consistent to keep showLogoModal()/hideLogoModal() and routing on the screen, even if it means some duplication. Also, at the very least, I believe onVerifySuccess should be moved to the screen. The reason is that router.dismissAll() and router.back() are functions that should only be called within a specific navigation state. In other situations, different dismiss logic would be necessary.